pkg/util/leaderlease/leaderlease.go

```diff
@@ -0,0 +1,269 @@
+	value  string
+	ttl    uint64
+
+	// the fraction of the ttl to wait before trying to renew
+	waitFraction float32
+	// the interval to wait when an error occurs acquiring the lease
+	pauseInterval time.Duration
+	// the maximum retries to renew the lease
+	renewRetries int
+	// the shortest time between attempts to renew the lease
+	minimumRetryInterval time.Duration
+}
+
+// NewEtcd creates a Lease in etcd, storing value at key with expiration ttl
+// and continues to refresh it until the key is deleted, expires, or another
+// client takes it.
+func NewEtcd(client *etcdclient.Client, key, value string, ttl uint64) Leaser {
+	return &Etcd{
+		client: client,
+		key:    key,
+		value:  value,
+		ttl:    ttl,
+
+		waitFraction:         0.75,
+		pauseInterval:        time.Second,
+		renewRetries:         10,
+		minimumRetryInterval: 100 * time.Millisecond,
+	}
+}
+
+// AcquireAndHold implements an acquire and release of a lease.
+func (e *Etcd) AcquireAndHold(notify chan struct{}) {
+	for {
+		ok, ttl, index, err := e.tryAcquire()
+		if err != nil {
+			util.HandleError(err)
+			time.Sleep(e.pauseInterval)
+			continue
+		}
+		if !ok {
+			time.Sleep(e.pauseInterval)
+			continue
+		}
+
+		// notify
+		notify <- struct{}{}
+		defer close(notify)
+
+		// hold the lease
+		if err := e.tryHold(ttl, index); err != nil {
+			util.HandleError(err)
+		}
+		break
+	}
+}
+
+// tryAcquire tries to create the lease key in etcd, or if it already exists
+// and belongs to another user, to wait until the lease expires or is deleted.
+func (e *Etcd) tryAcquire() (ok bool, ttl uint64, index uint64, err error) {
+	ttl = e.ttl
+
+	resp, err := e.client.Create(e.key, e.value, ttl)
+	if err == nil {
+		// we hold the lease
+		glog.V(4).Infof("Lease %s acquired, ttl %d seconds", e.key, e.ttl)
+		return true, ttl, resp.EtcdIndex + 1, nil
+	}
+
+	if !storage.IsEtcdNodeExist(err) {
+		return false, 0, 0, fmt.Errorf("unable to check lease %s: %v", e.key, err)
+	}
+
+	latest, err := e.client.Get(e.key, false, false)
+	if err != nil {
+		return false, 0, 0, fmt.Errorf("unable to retrieve lease %s: %v", e.key, err)
+	}
+
+	index = eventIndexFor(latest)
+	if latest.Node.TTL > 0 {
+		ttl = uint64(latest.Node.TTL)
+	}
+
+	if latest.Node.Value != e.value {
+		glog.V(4).Infof("Lease %s owned by %s, waiting for expiration", e.key, latest.Node.Value)
+		// waits until the lease expires or changes to us
+		if err := e.waitForExpiration(false, &index, nil); err != nil {
+			return false, 0, 0, fmt.Errorf("unable to wait for lease expiration %s: %v", e.key, err)
+		}
+		return false, 0, 0, nil
+	}
+
+	glog.V(4).Infof("Lease %s already held, expires in %d seconds", e.key, ttl)
+
+	return true, ttl, index, nil
+}
+
+// Release tries to delete the leader lock.
+func (e *Etcd) Release() {
+	for i := 0; i < e.renewRetries; i++ {
+		_, err := e.client.CompareAndDelete(e.key, e.value, 0)
+		if err == nil {
+			break
+		}
+		if storage.IsEtcdTestFailed(err) || storage.IsEtcdNotFound(err) {
```

Comment:
If test failed, someone else holds the lock. If Not Found, the lock has
been deleted.

On Fri, Aug 14, 2015 at 4:37 PM, David Eads <notifications@github.com>
wrote:

> In pkg/util/leaderlease/leaderlease.go
> <https://github.com/openshift/origin/pull/4137#discussion_r37116796>:
>
> > +		return false, 0, 0, nil
> > +	}
> > +
> > +	glog.V(4).Infof("Lease %s already held, expires in %d seconds", e.key, ttl)
> > +
> > +	return true, ttl, index, nil
> > +}
> > +
> > +// Release tries to delete the leader lock.
> > +func (e *Etcd) Release() {
> > +	for i := 0; i < e.renewRetries; i++ {
> > +		_, err := e.client.CompareAndDelete(e.key, e.value, 0)
> > +		if err == nil {
> > +			break
> > +		}
> > +		if storage.IsEtcdTestFailed(err) || storage.IsEtcdNotFound(err) {
>
> Why are these errors whitelisted?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4137/files#r37116796>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift